nn/loss.py

```diff
@@ -57,18 +57,32 @@ def cross_entropy(*, target: nn.Tensor, estimated: nn.Tensor, estimated_type: st
   return -nn.dot(target, log_prob, reduce=axis)
 
 
-def kl_div(*, target: nn.Tensor, estimated: nn.Tensor):
+def kl_div(*, target: nn.Tensor, estimated: nn.Tensor, estimated_type: str, axis: Optional[nn.Dim] = None) -> nn.Tensor:
```

Comment:
I thought about this too. Only log space would not work here, since we need both log and probs. Only check would maybe be whether logits or probs. 
This isn't done for `ce` either, so maybe we should leave it like this here and maybe have one common PR change it?

Reply:
Why does log space not work? You can simply do `exp(log_prob)` to get the probs?

But you are right, for CE we also don't have it. So maybe it's fine for now and this can be an extension later whenever we need it.

Note however that it can be inefficient this way because it might compute softmax twice. The `safe_log` on a previous softmax will convert it into a `log_softmax`, so then it calculates both the softmax and the log_softmax. It would be more efficient to only calculate log_softmax and then do just `exp` on the log probs.

Reply:
>You can simply do exp(log_prob) to get the probs?

If we assume that its always log to base e yes. I guess other bases are too uncommon to be used so its fine?

Reply:
>It would be more efficient to only calculate log_softmax and then do just exp on the log probs.

So you mean like this: `kl = nn.exp(log_target) * (log_target - log_est)` ? Or where would be calc. softmax twice?

Reply:
> > It would be more efficient to only calculate log_softmax and then do just exp on the log probs.
> 
> So you mean like this: `kl = nn.exp(log_target) * (log_target - log_est)` ? Or where would be calc. softmax twice?

In `kl = target * (nn.safe_log(target) - log_est)`, we would calc softmax twice for target: `target = softmax(...)` and `nn.safe_log(target) = log_softmax(...)`.

Reply:
Well, I'm not sure what's better (from a stability point of view). Doing exp(log_softmax(...)) vs softmax(...).

Reply:
Also, despite "better" (both under efficiency consideration, and stability), what's more "straightforward", i.e. what would the user expect what would happen here?

Reply:
So, maybe leave it like it is for now?

Reply:
As long as its fine in Efficiency, I think as it is its more readable and closer to the formula. So if you say its fine, I would leave it like this.

Reply:
Well, as said, it's a bit suboptimal wrt efficiency. The current code results in softmax twice. The question is, what is more important, efficiency or maybe straightforwardness?

Maybe efficiency is actually more important because that is in general for RETURNN the policy.

Reply:
See my suggested code change.